src/core/lib/Bombe.mjs

```diff
@@ -0,0 +1,472 @@
+     * @returns {string[][2]} - list of pairs of candidate rotor setting, and calculated stecker pair
+     */
+    run() {
+        let stops = 0;
+        const result = [];
+        // For each possible rotor setting
+        const nChecks = Math.pow(26, this.baseRotors.length);
+        for (let i=1; i<=nChecks; i++) {
+            this.wires.fill(false);
+            // Energise the test input, follow the current through each scrambler
+            // (and the diagonal board)
+            this.energiseCount = 0;
+            this.energise(this.testInput);
+            // Count the energised outputs
+            let count = 0;
+            for (let j=26*this.testRegister; j<26*(1+this.testRegister); j++) {
+                if (this.wires[j]) {
+                    count++;
+                }
+            }
+            // If it's not all of them, we have a stop
+            if (count < 26) {
+                stops += 1;
+                let stecker;
+                // The Bombe tells us one stecker pair as well. The input wire and test register we
+                // started with are hypothesised to be a stecker pair.
+                if (count === 25) {
+                    // Our steckering hypothesis is wrong. Correct value is the un-energised wire.
+                    for (let j=0; j<26; j++) {
+                        if (!this.wires[26*this.testRegister + j]) {
+                            stecker = `${i2a(this.testRegister)} <-> ${i2a(j)}`;
+                            break;
+                        }
+                    }
+                } else if (count === 1) {
+                    // This means our hypothesis for the steckering is correct.
+                    stecker = `${i2a(this.testRegister)} <-> ${i2a(this.testInput[1])}`;
+                } else {
+                    // Unusual, probably indicative of a poor menu. I'm a little unclear on how
+                    // this was really handled, but we'll return it for the moment.
+                    stecker = `? (wire count: ${count})`;
+                }
+                result.push([this.indicator.getPos(), stecker]);
+            }
+            // Step all the scramblers
+            // This loop counts how many rotors have reached their starting position (meaning the
+            // next one needs to step as well)
+            let n = 1;
+            for (let j=1; j<this.baseRotors.length; j++) {
+                if ((i % Math.pow(26, j)) === 0) {
+                    n++;
+                } else {
+                    break;
+                }
+            }
+            for (const scrambler of this.allScramblers) {
+                scrambler.step(n);
+            }
+            // Send status messages at what seems to be a reasonably sensible frequency
+            if (n > 2) {
+                const msg = `Bombe run with ${this.nLoops} loops in menu (2+ desirable): ${stops} stops, ${Math.floor(100 * i / nChecks)}% done`;
+                this.update(msg);
```

Comment:
To be honest, we could reduce the frequency of this update message even further. It fires a lot at the moment.